### DISCUSSION TITLE
PyMuPDF extraction

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#3791] [Looking for help]
Can PyMuPDF extract PDF with this text
[microsoft.pdf](https://github.com/user-attachments/files/16657168/microsoft.pdf)

**Comment by JorjMcKie** [comment]
What kept you keep away from trying this yourself:

![image](https://github.com/user-attachments/assets/6a9fe90a-b276-40ce-b37b-e7897e65e7ff)

**Reply by meghanaviyyapu** [reply]
Thanks I think I missed something. Got a different output earlier. I have tried extracting from the below PDF(page 4) but the extracted text looks different from displayed text
![image](https://github.com/user-attachments/assets/045dcc31-e1f4-466a-80ba-1d922ec04cbf)
[Telugu Indic Input 3-User Guide.pdf](https://github.com/user-attachments/files/16659072/Telugu.Indic.Input.3-User.Guide.pdf)

**Reply by JorjMcKie** [reply]
Yes, but that is true for all / multiple text extractors ...  Adobe among them.
The major thing seems to be that "o" symbol, which is always replaced by a space.

**Reply by meghanaviyyapu** [reply]
Ok is there a way to extract the text in the same way as in input PDF?

**Reply by JorjMcKie** [reply]
I'm afraid not without using OCR ...

**Reply by meghanaviyyapu** [reply]
Ok thanks

**Reply by meghanaviyyapu** [reply]
Can it be possible using PyMuPDF?

**Reply by JorjMcKie** [reply]
Yes - did you notice the integrated OCR feature?
Possibly however, your installed Tesseract app needs to be extended by adding supported languages to it.

**Reply by meghanaviyyapu** [reply]
I am yet to go through the OCR feature. Will try the same, thanks

**Reply by meghanaviyyapu** [reply]
May I know if you have tried Adobe extraction before as you had mentioned that other text extractors have a similar issue?

**Reply by JorjMcKie** [reply]
Yes, I did. Also Nitro PDF and pdftotext from XPDF.

**Reply by meghanaviyyapu** [reply]
Ok thanks for the info

**Reply by meghanaviyyapu** [reply]
I have tried extracting the following PDF. The text in different languages got extracted pretty well. May I know why PyMuPDF is able to extract the same language in some PDFs and not able to accurately extract few characters in some? How to identify which characters are extractable and which are not?
[ocr.pdf](https://github.com/user-attachments/files/16689576/ocr.pdf)

**Reply by JorjMcKie** [reply]
As I told you: other extractors have the same problem.
Whether or not some characters can be extracted can have have dozens of different reasons - including errors committed by the PDF or font creator.
If characters cannot be extracted, we cannot know them, right? We also cannot predict anything of course. So this is not a logical question.
Also, that symbol looking like an "o" may still be encoded by 0x20 (Unicode 32) in the font. This is not illegal. And 0x20 is space in all text extractions.

**Reply by meghanaviyyapu** [reply]
In the above ocr.pdf the text got extracted accurately. Is it because of no error in the PDF or font?

**Reply by JorjMcKie** [reply]
OCR engines try to match image content with characters they know about. Every OCR will of course recognize what "o" means and return an o (the Latin character).
This does not mean that this is the correct, intended original character encoded by the PDF creator!
OCR engines do not semantically **_understand_** what they are extracting!!!

**Reply by JorjMcKie** [reply]
If an OCR detects some dirty spot on a page, it will try to match one of the characters it knows about. It cannot understand what it is looking at.

**Reply by meghanaviyyapu** [reply]
I haven't used ocr to extract the above text. I have used PyMuPDF page.get_text() for the PDF "ocr.pdf" and the text extraction was the same as in input PDF whereas for the PDF uploaded yesterday o was missing. I understand that it is difficult to identify which characters can be extracted and which cannot. How can it be decided when to use ocr for text based multi lingual content PDFs?

**Reply by JorjMcKie** [reply]
Trial an error!

Please also accept that while OCR-ing may produce an acceptably looking result, it may still not produce the intended original text - see above example of that "o" symbol: if an OCR identifies it as a Latin character then this may look good, but is still wrong compared to the original text's meaning.

**Reply by meghanaviyyapu** [reply]
May I know if you are referring to the below?
![image](https://github.com/user-attachments/assets/742e0288-b433-4bc5-80d0-b42afbd9822d)
